### PR TITLE
Cleanup IP fragmentation, TCP session and TLS sessions

### PR DESCRIPTION
**Pull request by gpotter2**
- all `Sessions`: replace `on_received_packet` by two functions: `recv` that is an iterator and `process`. This makes a cleaner separation, and re-puts the storing/prn/etc; back into `sniff` (where it should have been left)
- cleanup `TCPSession`:
  - now do not dissect the packet just to re-build it afterwards, thanks to a new `stop_payload_dissection` attribute, which allows the caller to say "Don't parse above TCP". This required changing the layout of `recv()` in SuperSocket to allow for extra arguments.
  - better handling of out-of-order packets: handle the case where the first packet received wasn't the first (fixes https://github.com/secdev/scapy/issues/4030)
- cleanup IP fragmentation utils:
  - deduplicate the two `fragment()` functions
  - fix https://github.com/secdev/scapy/pull/3970
- cleanup TLS sessions
  - fix TLS sessions being badly recognized when over TCP defragmented sessions (with TCPSession)
  - fix `repr()` not showing the correct direction in tls sessions
  - deprecate `TLSSession` (literally brings nothing. Just use `TCPSession` with `conf.tls_session_enable=True`). This is now also fixed thanks to the above fix.
  - fix the handling of NSS files: now supports multiple streams (fixes https://github.com/secdev/scapy/pull/3374#issuecomment-981634748)
  - there was a bug in our handling of extended master secret key that made it so that we always called `compute_ms_and_derive_keys` even if no `pre_master_secret_key` was computed (i.e. we don't have any private key). This is now fixed, the function is only called whenever the `pre_master_secret_key` has been computed.
  - fix https://github.com/secdev/scapy/issues/3722 (the NSS keys were only computed when the bug above occured, leading to sessions not using extended master key to fail to compute them, and making the bug unclear). This is now properly fixed: NSS keys are checked separately as last resort.

fixes https://github.com/secdev/scapy/issues/4030
fixes https://github.com/secdev/scapy/issues/3722
closes https://github.com/secdev/scapy/pull/3966
closes https://github.com/secdev/scapy/pull/3970

### CODE REVIEW

**Comment by stulle123** [comment]
Hi,

When I run this PR on my [test pcap](https://github.com/secdev/scapy/files/12231501/test.pcap.zip) I'm getting:

```bash
(venv) $ python3 test.py
WARNING: No IPv4 address found on anpi1 !
WARNING: No IPv4 address found on anpi2 !
WARNING: more No IPv4 address found on anpi0 !
WARNING: TLSSession is deprecated and will be removed in a future version. Please use TCPSession instead with conf.tls_session_enable=True
Traceback (most recent call last):
  File "test.py", line 9, in <module>
    capture = sniff(offline="test.pcap", session=TCPSession)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/scapy/sendrecv.py", line 1311, in sniff
    sniffer._run(*args, **kwargs)
  File "/venv/lib/python3.11/site-packages/scapy/sendrecv.py", line 1254, in _run
    session.on_packet_received(p)
  File "/venv/lib/python3.11/site-packages/scapy/sessions.py", line 396, in on_packet_received
    pkt = self._process_packet(pkt)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/scapy/sessions.py", line 359, in _process_packet
    packet = tcp_reassemble(
             ^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/scapy/layers/tls/session.py", line 1128, in tcp_reassemble
    return cls(data, _underlayer=underlayer)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/scapy/base_classes.py", line 398, in __call__
    i.__init__(*args, **kargs)
  File "/venv/lib/python3.11/site-packages/scapy/layers/tls/record.py", line 303, in __init__
    super(TLS, self).__init__(*args, **kargs)
  File "/venv/lib/python3.11/site-packages/scapy/layers/tls/session.py", line 995, in __init__
    Packet.__init__(self, _pkt=_pkt, post_transform=post_transform,
  File "/venv/lib/python3.11/site-packages/scapy/packet.py", line 183, in __init__
    self.dissect(_pkt)
  File "/venv/lib/python3.11/site-packages/scapy/packet.py", line 1070, in dissect
    s = self.pre_dissect(s)
        ^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/scapy/layers/tls/record.py", line 528, in pre_dissect
    iv, cfrag, mac = self._tls_auth_decrypt(hdr, efrag)
    ^^^^^^^^^^^^^^
ValueError: not enough values to unpack (expected 3, got 2)
```

Code:

```python
from scapy.config import conf
from scapy.main import load_layer
from scapy.sendrecv import sniff
from scapy.sessions import TCPSession

conf.tls_session_enable=True
load_layer("tls")

capture = sniff(offline="test.pcap", session=TCPSession)
```

Cheers

File: scapy/layers/inet.py [hunk 1]
```diff
@@ -1181,86 +1152,112 @@ def overlap_frag(p, overlap, fragsize=8, overlap_fragsize=None):
...
     It contains the logic behind the defrag() and defragment() functions
     """
-    frags = defaultdict(lambda: [])
+    frags = defaultdict(list)
     final = []
-    pos = 0
-    for p in plist:
-        p._defrag_pos = pos
-        pos += 1
-        if IP in p:
```
Comment (evverx) [inline scapy/layers/inet.py:1234]: Looks like without this check `sniff` fails with
```python
Traceback (most recent call last):
    sniff(offline='pcap-with-arp-stuff.pcap', session=IPSession)
  File "scapy/sendrecv.py", line 1311, in sniff
    sniffer._run(*args, **kwargs)
  File "scapy/scapy/sendrecv.py", line 1254, in _run
    session.on_packet_received(p)
  File "scapy/scapy/sessions.py", line 144, in on_packet_received
    super(IPSession, self).on_packet_received(self._ip_process_packet(pkt))
                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "scapy/scapy/sessions.py", line 138, in _ip_process_packet
    return _defrag_ip_pkt(packet, self.fragments)[1]  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "scapy/scapy/layers/inet.py", line 1183, in _defrag_ip_pkt
    ip = pkt[IP]
         ~~~^^^^
  File "scapy/scapy/packet.py", line 1370, in __getitem__
    raise IndexError("Layer [%s] not found" % name)
IndexError: Layer [IP] not found
```
when pcaps contain at least one ARP packet (or any other packet without the IP layer).

Comment (gpotter2) [reply scapy/layers/inet.py:1234]: Good catch, thanks !

**Comment by dschmidt-ae** [comment]
Hi,

Just ran another test with this [pcap](https://github.com/secdev/scapy/files/12300645/test_2.pcap.zip) and `tcp_reassemble()` (in `scapy/layers/tls/session.py`) takes forever to process.

```python
from scapy.config import conf
from scapy.main import load_layer
from scapy.sendrecv import sniff
from scapy.sessions import TCPSession

conf.tls_session_enable=True
load_layer("tls")


capture = sniff(offline="test_2.pcap", session=TCPSession)

print("Done")
```

Pyinstrument output:

```bash
$ pyinstrument test.py -r text --show '*/scapy/*'    

  _     ._   __/__   _ _  _  _ _/_   Recorded: 11:34:08  Samples:  354062
 /_//_/// /_\ / //_// / //_'/ //     Duration: 361.988   CPU time: 360.251
/   _/                      v4.5.1

Program: test.py

361.987 <module>  test.py:1
└─ 359.257 sniff  scapy/sendrecv.py:1307
   └─ 359.257 AsyncSniffer._run  scapy/sendrecv.py:1064
      └─ 358.970 TCPSession.on_packet_received  scapy/sessions.py:385
         └─ 358.947 TCPSession._process_packet  scapy/sessions.py:288
            └─ 358.330 TLS.tcp_reassemble  scapy/layers/tls/session.py:1124
               └─ 353.972 [self]  scapy/layers/tls/session.py
```

**Comment by stulle123** [comment]
Hi,

Just ran another test with this [pcap](https://github.com/secdev/scapy/files/12300645/test_2.pcap.zip) and `tcp_reassemble()` (in `scapy/layers/tls/session.py`) takes forever to process.

```python
from scapy.config import conf
from scapy.main import load_layer
from scapy.sendrecv import sniff
from scapy.sessions import TCPSession

conf.tls_session_enable=True
load_layer("tls")


capture = sniff(offline="test_2.pcap", session=TCPSession)

print("Done")
```

Pyinstrument output:

```bash
$ pyinstrument test.py -r text --show '*/scapy/*'    

  _     ._   __/__   _ _  _  _ _/_   Recorded: 11:34:08  Samples:  354062
 /_//_/// /_\ / //_// / //_'/ //     Duration: 361.988   CPU time: 360.251
/   _/                      v4.5.1

Program: test.py

361.987 <module>  test.py:1
└─ 359.257 sniff  scapy/sendrecv.py:1307
   └─ 359.257 AsyncSniffer._run  scapy/sendrecv.py:1064
      └─ 358.970 TCPSession.on_packet_received  scapy/sessions.py:385
         └─ 358.947 TCPSession._process_packet  scapy/sessions.py:288
            └─ 358.330 TLS.tcp_reassemble  scapy/layers/tls/session.py:1124
               └─ 353.972 [self]  scapy/layers/tls/session.py
```

**Comment by gpotter2** [comment]
Thanks for the report, I'll have a look.

**Comment by stulle123** [comment]
Hey @gpotter2,

Sorry for bothering :-D found another issue when processing this [pcap](https://github.com/secdev/scapy/files/12357616/chunk.zip) (using the same code from above):

```bash
WARNING: No IPv4 address found on anpi2 !
WARNING: No IPv4 address found on anpi1 !
WARNING: more No IPv4 address found on anpi0 !
WARNING: Socket <scapy.utils.PcapNgReader object at 0x110de25d0> failed with 'PcapNgReader.recv() got an unexpected keyword argument 'stop_payload_dissection''. It was closed.
```

Cheers

**Comment by gpotter2** [comment]
Thanks. I'm currently rewriting the PR quite a bit.

**Comment by gpotter2** [comment]
@stulle123 this is fixed thanks.

**Comment by polybassa** [comment]
regarding the failing test, somehow the behavior of the `count` parameter changed. 
```
diff --git a/test/contrib/automotive/ecu.uts b/test/contrib/automotive/ecu.uts
index 680a2b33..4ced5203 100644
--- a/test/contrib/automotive/ecu.uts
+++ b/test/contrib/automotive/ecu.uts
@@ -673,7 +673,7 @@ with PcapReader(scapy_path("test/pcaps/gmlan_trace.pcap.gz")) as sock:
     print("Check 1 after change to diagnostic mode")
     assert len(ecu.supported_responses) == 1
     assert ecu.state == EcuState(session=3)
-    gmlanmsgs = sniff(session=ISOTPSession(supersession=session, rx_id=[0x241, 0x641, 0x101], basecls=GMLAN), count=8, opened_socket=sock)
+    gmlanmsgs = sniff(session=ISOTPSession(supersession=session, rx_id=[0x241, 0x641, 0x101], basecls=GMLAN), count=6, opened_socket=sock)
     ecu = session.ecu
     print("Check 2 after some more messages were read1")
     assert len(ecu.supported_responses) == 3
@@ -681,13 +681,13 @@ with PcapReader(scapy_path("test/pcaps/gmlan_trace.pcap.gz")) as sock:
     assert ecu.state.session == 3
     print("assert 1")
     assert ecu.state.communication_control == 1
-    gmlanmsgs = sniff(session=ISOTPSession(supersession=session, rx_id=[0x241, 0x641, 0x101], basecls=GMLAN), count=10, opened_socket=sock)
+    gmlanmsgs = sniff(session=ISOTPSession(supersession=session, rx_id=[0x241, 0x641, 0x101], basecls=GMLAN), count=2, opened_socket=sock)
     ecu = session.ecu
     print("Check 3 after change to programming mode (bootloader)")
     assert len(ecu.supported_responses) == 4
     assert ecu.state.session == 2
     assert ecu.state.communication_control == 1
-    gmlanmsgs = sniff(session=ISOTPSession(supersession=session, rx_id=[0x241, 0x641, 0x101], basecls=GMLAN), count=16, opened_socket=sock)
+    gmlanmsgs = sniff(session=ISOTPSession(supersession=session, rx_id=[0x241, 0x641, 0x101], basecls=GMLAN), count=6, opened_socket=sock)
     ecu = session.ecu
     print("Check 4 after gaining security access")
     assert len(ecu.supported_responses) == 6

```

This changes make the test work again

**Comment by gpotter2** [comment]
Thanks a lot !!!

**Comment by polybassa** [comment]
I would say, it's more "clean / correct" now. In the past, the session object itself stored the packets and count was evaluated against that. Therefore I needed to place the "total" number of packets, stored in the session, into the count parameter of sniff.  Now you place the number of packet you want to receive there.

**Review by polybassa — DISMISSED** [review]

**Comment by evverx** [comment]
With this PR applied `sniff` seems to fail with
```
WARNING: Socket <scapy.arch.bpf.supersocket.L2bpfListenSocket object at 0x10e4e5910> failed with ''Conf' object has no attribute 'debug_dissect''. It was closed.
```
from time to time. I'll try to collect packets triggering it and attach a pcap.

**Comment by evverx** [comment]
Looks like it has nothing to do with this PR. It can be fixed with:
```diff
diff --git a/scapy/layers/tls/record_sslv2.py b/scapy/layers/tls/record_sslv2.py
index 8d311faa..8e99a352 100644
--- a/scapy/layers/tls/record_sslv2.py
+++ b/scapy/layers/tls/record_sslv2.py
@@ -174,7 +174,7 @@ class SSLv2(TLS):
             except KeyboardInterrupt:
                 raise
             except Exception:
-                if conf.debug_dissect:
+                if conf.debug_dissector:
                     raise
                 p = conf.raw_layer(s, _internal=1, _underlayer=self)
             self.add_payload(p)
```

**Comment by gpotter2** [comment]
Still waiting for more review !

**Review by p-l- — COMMENTED** [review]
I don't have enough time for now to do a comprehensive review but overall LGTM, great work!

**Review by guedou — CHANGES_REQUESTED** [review]
LGTM I will try to test the code further.

**Pull request merged**